src/Ouroboros/Network/Protocol/Untyped.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeInType #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE GADTs #-}
+
+-- | An view of the typed protocol where we forget the strong typing on the
+-- protocol state transitions and just retain the protocol actions.
+--
+-- This view is useful for executing protocols.
+--
+module Ouroboros.Network.Protocol.Untyped where
+
+import qualified Ouroboros.Network.Protocol.Typed as Typed
+
+import Data.Proxy
+import Data.Typeable (Typeable, (:~:)(Refl), eqT)
+import Data.Kind (Type)
+
+
+-- | This is like 'Typed.Peer' but without the strong typing on the protocol
+-- state transitions.
+--
+-- Because it is not strongly typed, the 'PeerAwait' can fail if given an
+-- unexpected message that is not correct for the protocol state.
+--
+-- This form can be used for interpreting against a channel (message or byte
+-- based with a suitable codec).
+--
+data Peer msg m a
+  = PeerDone   a
+  | PeerHole  (m (Peer msg m a))
+  | PeerYield (SomeMessage msg) (Peer msg m a)
+  | PeerAwait (SomeMessage msg -> Maybe (Peer msg m a))
+
+
+asUntypedPeer :: forall p st
+                        (msg     :: st -> st -> Type)
+                        (statusf :: st -> Typed.Status st)
+                        (statust :: st -> Typed.Status st)
+                        (from    :: st)
+                        (to      :: st)
+                        m a .
+                 Monad m
+              => Typed.Peer p msg (statusf from) (statust to) m a
+              -> Peer msg m a
+
+asUntypedPeer (Typed.PeerDone a) = PeerDone a
+
+asUntypedPeer (Typed.PeerHole f) = PeerHole (fmap asUntypedPeer f)
+
+asUntypedPeer (Typed.PeerYield msg k) =
+    PeerYield (SomeMessage (Typed.exchangeTransition msg))
+              (asUntypedPeer k)
+
+asUntypedPeer (Typed.PeerAwait k) =
+    PeerAwait $ \amsg ->
+      case castTransition (Proxy :: Proxy from) amsg of
+        Unexpected   -> Nothing
+        Expected msg -> Just (asUntypedPeer (k msg))
+
+
+data SomeMessage (msg :: st -> st -> Type) where
+  -- | Must ensure Typeable here, because we can't do it at 'castTransition'.
+  -- That would require stating that every type in the state kind is
+  -- typeable. Quantified constraints could help, although a short attempt at
+  -- that did not work.
+  SomeMessage :: Typeable from => msg from to -> SomeMessage msg
+
+
+data TransitionFrom msg from where
+  Expected   :: msg from to -> TransitionFrom msg from
+  Unexpected :: TransitionFrom msg from
+
+castTransition
+  :: forall st (msg :: st -> st -> Type) (from :: st) .
+     ( Typeable from )
+  => Proxy from
+  -> SomeMessage msg
+  -> TransitionFrom msg from
+castTransition _proxy (SomeMessage (m :: msg from' to)) =
```

Comment:
Naming inconsistency: `castTransition`, `TransitionFrom`, but `SomeMessage`. Are we calling them messages or transitions? I like "transition". "message" carries connotations from networking/IPC. There's nothing inherently IPC about the typed-transitions style. A typed-transitions presentation can be run in  a single process.

Reply:
Yep fair point. Will review names there.

I'm trying to use the name message, precisely becaue this is about message passing and I think it helps with comprehension. You're right of course that it's in principle more general, but that's not how we're using it here. We're using it for protocols (which usually but don't have to be between concurrent processes).

Indeed I plan to rename Hole -> Effect, and Yield/Await to Send/Recv. I relaise it pains you to see us specialise the naming to the protocol/IPC use case, but I think it's worth it to make it all more comprehensible. It was honestly a real slog at first (which is partly why it was so long before proper review). The naming has something to do with that.